src/coreclr/src/jit/morph.cpp

```diff
@@ -12515,25 +12515,49 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op2 = tree->AsOp()->gtOp2;
             }
 
-            /* See if we can fold GT_ADD nodes. */
-
-            if (oper == GT_ADD)
+            // See if we can fold constants for commutative operators.
+            if (tree->OperIs(GT_OR, GT_XOR, GT_AND, GT_ADD))
             {
-                /* Fold "((x+icon1)+(y+icon2)) to ((x+y)+(icon1+icon2))" */
-
-                if (op1->gtOper == GT_ADD && op2->gtOper == GT_ADD && !gtIsActiveCSE_Candidate(op2) &&
-                    op1->AsOp()->gtOp2->gtOper == GT_CNS_INT && op2->AsOp()->gtOp2->gtOper == GT_CNS_INT &&
-                    !op1->gtOverflow() && !op2->gtOverflow())
+                // Fold "((x <op> icon1) <op> (y <op> icon2))" to "((x <op> y) <op> (icon1 <op> icon2))"
+                if (op1->OperIs(op2->OperGet()) && op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op2) &&
+                    op1->AsOp()->gtGetOp2()->IsCnsIntOrI() && op2->AsOp()->gtGetOp2()->IsCnsIntOrI())
                 {
                     // Don't create a byref pointer that may point outside of the ref object.
                     // If a GC happens, the byref won't get updated. This can happen if one
                     // of the int components is negative. It also requires the address generation
```

Comment:
>// It also requires the address generation be in a fully-interruptible code region.

That comment looks incorrect, at least I do not see where we check that ` a fully-interruptible code region` condition.

It is also not clear for me in which case that optimization for `varTypeIsGC` is incorrect, could somebody explain that to me please?